tests/list-test.cc

```diff
@@ -0,0 +1,211 @@
+// Copyright Microsoft and CHERIoT Contributors.
+// SPDX-License-Identifier: MIT
+
+#define TEST_NAME "List"
+#include "tests.hh"
+#include <ds/linked_list.h>
+
+namespace
+{
+	using ObjectRing = ds::linked_list::cell::Pointer;
+
+	/**
+	 * Example class we want to link into a doubly linked list.
+	 *
+	 * The class contains a single integer for the purposes of the test.
+	 *
+	 * `ds::linked_list` is an intrusive list: we embed the list node into
+	 * the class we want to link.  There are various implementations of the
+	 * list nodes. Here we use the most simple one
+	 * (`ds::linked_list::cell::Pointer`) which relies on two pointers
+	 * `next` and `prev`.
+	 */
+	struct LinkedObject
+	{
+		int data;
+		/**
+		 * List node: links objects into the doubly-linked list.
+		 */
+		ObjectRing ring __attribute__((__cheri_no_subobject_bounds__)) = {};
+		/**
+		 * Container-of for the above field. This is used to retrieve the
+		 * corresponding object from a list element.
+		 */
+		__always_inline static struct LinkedObject *from_ring(ObjectRing *c)
+		{
+			return reinterpret_cast<struct LinkedObject *>(
+			  reinterpret_cast<uintptr_t>(c) -
+			  offsetof(struct LinkedObject, ring));
+		}
+	};
+} // namespace
+
+/**
+ * The sentinel is a special class encapsulating a list node. It provides many
+ * handy functions to operate on the list. The sentinel is thus part of the
+ * list, without being one of the objects linked.
+ *
+ * Note: do not attempt to define the sentinel on the stack, because it would
+ * lead some list nodes holding a pointer to a stack value, i.e., to an invalid
+ * capability. This would lead to a crash while using the list.
+ */
+ds::linked_list::Sentinel<ObjectRing> objects = {};
+
+void test_list()
+{
+	debug_log("Testing the list implementation.");
+
+	// Number of elements we will add to the list in the test. Must be
+	// divisible by two.
+	static constexpr int NumberOfListElements = 30;
+
+	auto heapAtStart = heap_quota_remaining(MALLOC_CAPABILITY);
+
+	TEST(objects.is_empty(), "Newly created list is not empty");
+
+	// Create heap-allocated objects, and link them into the linked list.
+	for (int i = 0; i < NumberOfListElements; i++)
+	{
+		Timeout       t{UnlimitedTimeout};
+		LinkedObject *o = (LinkedObject *)heap_allocate(
+		  &t, MALLOC_CAPABILITY, sizeof(LinkedObject));
+		TEST(CHERI::Capability{o}.is_valid(), "Cannot allocate linked object");
+
+		// Use the object integer as an index.
+		o->data = i;
+		// The list node has not yet been initialized.
+		o->ring.cell_reset();
+
+		// Test that we can retrieve the object from the link node and
+		// that this results in a valid capability.
+		TEST(o == LinkedObject::from_ring(&(o->ring)),
```

Comment:
`==` on pointers in CHERI C/C++ presently defaults to address-only equality.  You might use `__builtin_cheri_equal_exact`, or wrap with `Capability` first (since its `operator==` uses that primitive), to be more stringent.

Reply:
Oh, great point! I will apply this to other comparisons in the test.